#1132 Assignment fixes

CarlosNZ opened the pull request:
Fix #1132

Much simpler now -- one hook with a couple of small helpers and one mutation to handle ALL assigning/reassigning/unassigning cases,  replacing two hooks and three mutations, and the logic is less complicated. 

Have tried many combinations of assigning, unassigning and re-assigning, including swapping, reassigning from multiple assignments at once, etc., and all seem to be working nicely.

Also tidied up the error handling a bit, since there was (for example), "setAssignmentError" statements that weren't actually being handled. Now they're all consolidated under one "assignment error". I don't love the UI when the error shows, but it's better than nothing (and hopefully something the user will rarely see). (I was going to use the "useToast" hook to display the error, but I realised there are [some problems with that that need fixing](https://github.com/openmsupply/application-manager-web-app/issues/1134) -- but once we've done that we can use Toast notifications to display errors.)

Also -- I've kept the unassignment behaviour the same (unassign _all_ sections from reviewer at once), but it can easily be the other way (one at a time, like assignment), and have left some commented-out code that can easily be swapped out to do it the other way. What do we think is best? (Ideally, it would be user-selectable, but I couldn't really think of a nice way to handle it in the UI without it becoming messy and cluttered, so we can perhaps re-visit this at some point).

nmadruga commented on src/utils/hooks/useUpdateAssignment.tsx:
```diff
@@ -0,0 +1,86 @@
+  assignment: AssignmentDetails,
+  assignedSections: SectionAssignee | null
+) => {
+  if (!assignedSections) return { status: ReviewAssignmentStatus.Available }
```
Is this one to unassign a complete review_assignment? And is it only setting the status to `Available` since the back-end will make sure to set the array to empty/null?

CarlosNZ replied:
This line is just a guard for the case when "assignedSections" is null. Originally I had it that assignedSections couldn't be `null`, but when I was doing the unassignment, I needed a way to unassign from all sections without having to add a bunch of additional logic. 

If you look at "AssignmentSectionRow", you'll see my commented out code that would unassign one section at a time. If we went that way, we wouldn't need this line here either.

nmadruga reviewed (commented):
Seems very siimple to read. II liked it a lot. Now testing...

nmadruga commented:
Running into 2 problems that might require a little investigation - to make sure it's not a regretion. But if possible (even if not a regression) fixed as part of this PR, since the work will be paused for further issues...

Problem one. After re-assigning and un-assigning for tests. Looking on the **Overview tab** shows that the re-assigner/unassigner is missing from the review_assignment probably:
![Screen Shot 2022-03-07 at 12 22 50 PM](https://user-images.githubusercontent.com/16461988/156947023-7ae57e36-43f9-49dd-9078-bc69d643cdb7.png)

Problem two. What cause is not clear. It just happened after re-assigning, unassigning and re-assigning for tests. Not the user **Samanta** have this option: only **Not assigned** to choose from the dropdown. I'll try to replicate 🤭  
![Screen Shot 2022-03-07 at 12 24 14 PM](https://user-images.githubusercontent.com/16461988/156947022-3ddbfb59-591a-40fa-91bf-12c557865c2e.png)

nmadruga commented:
> I'll try to replicate 🤭

Ok - Following up, this happens un-related to trying to re-assign or unassign. If you just logs in as Samanata or Senior Reviewer at the start you can see the problem. Not sure why its showing the assignment to these users before someone is assigned.

CarlosNZ commented:
> Ok - Following up, this happens un-related to trying to re-assign or unassign. If you just logs in as Samanata or Senior Reviewer at the start you can see the problem. Not sure why its showing the assignment to these users before someone is assigned.

Isn't this what we expect? "Samantha" doesn't have assigning permissions for this template, so she shouldn't be able to assign anyone. Are you saying that the Dropdown menu shouldn't even be there? What should be, can you remember?

nmadruga commented:
> Are you saying that the Dropdown menu shouldn't even be there?

Yes, that's the problem. It should be showing anything to a user that can be a reviewer but isn't assigned.

CarlosNZ commented:
> Problem one. After re-assigning and un-assigning for tests. Looking on the **Overview tab** shows that the re-assigner/unassigner is missing from the review_assignment probably: ![Screen Shot 2022-03-07 at 12 22 50 PM](https://user-images.githubusercontent.com/16461988/156947023-7ae57e36-43f9-49dd-9078-bc69d643cdb7.png)

Weird that you are getting parsing errors in the Timeline, I can't replicate that at all. Are you using a fresh snapshot and checked there isn't stale records in the activity log?
<img width="845" alt="Screen Shot 2022-03-07 at 12 52 04 PM" src="https://user-images.githubusercontent.com/5456533/156947772-8323af34-5d72-4ad7-8cda-364f3ad03d6e.png">

If you can replicate this, can you save a snapshot in the error state and send it to me?

CarlosNZ commented:
> Yes, that's the problem. It should be showing anything to a user that can be a reviewer but isn't assigned.

Hmm, if that's changed it probably changed while doing the review_question_assignment removal. Can you remember what logic/state that component uses to check whether it should show the Dropdown?

nmadruga commented:
> > Yes, that's the problem. It should be showing anything to a user that can be a reviewer but isn't assigned.
> 
> Hmm, if that's changed it probably changed while doing the review_question_assignment removal. Can you remember what logic/state that component uses to check whether it should show the Dropdown?

I'll have a look

CarlosNZ commented:
Okay, have added guard to not show Dropdown or Unassign link when user doesn't have permission: 625b57b5

Note, this now means that a user can't even unassign themselves if they don't have assigning permissions -- which I guess is okay?

nmadruga commented:
> Okay, have added guard to not show Dropdown or Unassign link when user doesn't have permission: [625b57b](https://github.com/openmsupply/application-manager-web-app/commit/625b57b58bd77388206e3453be5872ec8e00ad80)
> 
> Note, this now means that a user can't even unassign themselves if they don't have assigning permissions -- which I guess is okay?

Not sure... What do you meant by assigning permission? If they have been assigned to a permission on that template that is `selfAssignable` they have permission and should see the dropdown.

I'm noticing a poblem now after self-assigning the "selected" reviewer isn't showing (before I click Submit)

CarlosNZ commented:
> Not sure... What do you meant by assigning permission? If they have been assigned to a permission on that template that is `selfAssignable` they have permission and should see the dropdown.

Yes, that's right. But before we had it that anyone could *UN*assign themselves from _any_ assignment regardless of permissions.

> I'm noticing a poblem now after self-assigning the "selected" reviewer isn't showing (before I click Submit)

What do you mean, exactly? When self-assigning, it says "Yourself" as the selected option before you Submit:
<img width="909" alt="Screen Shot 2022-03-07 at 2 49 49 PM" src="https://user-images.githubusercontent.com/5456533/156954258-08552483-0f76-458a-b26d-5a295f93df25.png">
Is that not what you're seeing?

nmadruga commented:
> Is that not what you're seeing?

Exactly, I wasn't seen that! I'll try again later today : )

CarlosNZ commented:
> > Is that not what you're seeing?
> 
> Exactly, I wasn't seen that! I'll try again later today : )

Okay, the problem was that the "Not assigned" option was disappearing from the options list after making a selection, and my check was checking for more than one item in the list.

I've updated it so that "Not Assigned" is always in the options list, and changed the "isLastLevel" check to just look at the first assignment and not depend on who is selected: 6b8df28c

CarlosNZ commented:
Also -- I realised that the assigner_id wasn't being updated in review_assignments, which is what was causing some of the Timeline parsing errors.

I've updated it so that the assigner_id is set to whoever *last* changed the review_assignment. So if Supervisor A assigns three sections, then Supervisor B  comes along later and unassigns it, the "assigner_id" will be for Supervisor B. d8819c56

Even if Supervisor B re-assigns one section (and leaves the other two as is), the assigner_id will be updated with Supervisor B's id. I don't see any other option, since there's only one assigner_id field per review_assignment. (The activity log will keep track of changes though.)

@nmadruga does that sound okay? Any other potential problems with the assigner_id field?

nmadruga reviewed (approved):
It looks good to me now Carl. I still find one not-expected behaivour when the re-assign/unassign are showing up on the previous level assignment after you select another Reviewer to re-assign on the Consolidation. And things are a bit buggy if I try to actually re-assign the Reviewer of previous level as well. But I think we can just add another bug fix for the future - not a very common path anyway.
